### PR TITLE
front-proxy: add --authentication-{pass-on,drop}-groups

### PR DESCRIPTION
**Pull request by sttts**
We want to eventually disallow `system:masters` and maybe other system groups. I.e. through the front-proxy only "normal" users should connect or controller kcp-admins.

### CODE REVIEW

File: cmd/kcp-front-proxy/authentication/groups.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+// GroupFilter is a filter that filters out group that are not in the allowed groups,
+// and groups that are in the disallowed groups.
+type GroupFilter struct {
+	Authenticator authenticator.Request
+
+	AllowedGroups    sets.String
+	DisallowedGroups sets.String
+
+	AllowedGroupPrefixes    []string
+	DisallowedGroupPrefixes []string
```
Comment (s-urbaniak) [inline cmd/kcp-front-proxy/authentication/groups.go:43]: see above

File: cmd/kcp-front-proxy/authentication/groups.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// AddFlags delegates to ClientCertAuthenticationOptions
+func (c *Authentication) AddFlags(fs *pflag.FlagSet) {
+	c.ClientCert.AddFlags(fs)
+
+	fs.StringSliceVar(&c.AllowedGroups, "authentication-allowed-groups", c.AllowedGroups,
+		"Groups that are passed on to the shard. Empty matches all. \"prefix*\" matches "+
+			"all beginning with the given prefix.")
+	fs.StringSliceVar(&c.DisallowedGroups, "authentication-disallowed-groups", c.DisallowedGroups,
```
Comment (s-urbaniak) [inline cmd/kcp-front-proxy/authentication/groups.go:130]: see above, `authentication-deny-groups` seems more concise.

**Comment by s-urbaniak** [comment]
what happens if we specify the same in `allowed-groups` and `disallow-groups` ?

File: cmd/kcp-front-proxy/authentication/groups.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	if len(a.AllowedGroups) > 0 || len(a.AllowedGroupPrefixes) > 0 {
+	nextGroup:
+		for _, g := range groupsToPassOn.UnsortedList() {
+			if !a.AllowedGroups.Has(g) {
+				for _, prefix := range a.AllowedGroupPrefixes {
+					if strings.HasPrefix(g, prefix) {
+						goto nextGroup
+					}
+				}
+				groupsToPassOn.Delete(g)
```
Comment (s-urbaniak) [inline cmd/kcp-front-proxy/authentication/groups.go:71]: ok, so `allowedGroups` has precedence over `disallowGroups`. we should document this, both in gdoc an CLI help.

File: cmd/kcp-front-proxy/options/authentication.go [hunk 4]
```diff
@@ -49,12 +56,45 @@ func (c *Authentication) ApplyTo(authenticationInfo *genericapiserver.Authentica
...
+	if len(c.AllowedGroups) > 0 || len(c.DisallowedGroups) > 0 {
+		filter := &kcpauthentication.GroupFilter{
+			Authenticator:    authenticationInfo.Authenticator,
+			AllowedGroups:    sets.NewString(),
+			DisallowedGroups: sets.NewString(),
+		}
+		authenticationInfo.Authenticator = filter
+
+		for _, g := range c.AllowedGroups {
+			if strings.HasSuffix(g, "*") {
```
Comment (s-urbaniak) [inline cmd/kcp-front-proxy/options/authentication.go:70]: can't we use https://pkg.go.dev/path/filepath#Glob ?

File: cmd/kcp-front-proxy/authentication/groups.go (hunk 1 above)
Comment (sttts) [reply cmd/kcp-front-proxy/authentication/groups.go:43]: done

File: cmd/kcp-front-proxy/authentication/groups.go (hunk 2 above)
Comment (sttts) [reply cmd/kcp-front-proxy/authentication/groups.go:130]: done

File: cmd/kcp-front-proxy/authentication/groups.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+		authenticationInfo.Authenticator = x509.NewDynamic(clientCAProvider.VerifyOptions, x509.CommonNameUserConversion)
+	}
+	return nil
+}
+
+// AddFlags delegates to ClientCertAuthenticationOptions
+func (c *Authentication) AddFlags(fs *pflag.FlagSet) {
+	c.ClientCert.AddFlags(fs)
+
+	fs.StringSliceVar(&c.AllowedGroups, "authentication-allowed-groups", c.AllowedGroups,
```
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/kcp-front-proxy/authentication/groups.go:127]: done

File: cmd/kcp-front-proxy/authentication/groups.go (hunk 3 above)
Comment (sttts) [reply cmd/kcp-front-proxy/authentication/groups.go:71]: good question. Clarified: `Disallowed groups trump over allowed groups.`

File: cmd/kcp-front-proxy/options/authentication.go (hunk 4 above)
Comment (sttts) [reply cmd/kcp-front-proxy/options/authentication.go:70]: We could. But then I would rather choose regexps than non-standard golang globbing. Also note that Glob has special meaning for slashes.

**Comment by s-urbaniak** [comment]
pass-on/drop is even better agreed 👍

**Comment by s-urbaniak** [comment]
last nit-pick: misses unit tests 😬

**Pull request merged**